src/background/services/deduplicator.test.ts

```diff
@@ -0,0 +1,323 @@
+        cacheFnArgs: true,
+      });
+      const dedupedFn2 = deduplicatorService.dedupe(fn2, {
+        cacheFnArgs: true,
+      });
+      // create same signature functions, but with different arguments
+      const resultPromises = [
+        dedupedFn1(1, {
+          nested: {
+            array: [1, 2, 3],
+            obj: { key: 'arg1' },
+          },
+        }),
+        dedupedFn2(2, {
+          nested: {
+            array: [4, 5, 6],
+            obj: { key: 'arg2' },
+          },
+        }),
+      ];
+
+      jest.runAllTimers();
+
+      const [result1, result2] = await Promise.all(resultPromises);
+
+      expect(result1).toBe(returnValueFn1);
+      expect(result2).toBe(returnValueFn2);
+      expect(fn1).toHaveBeenCalledTimes(1);
+      expect(fn2).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('Error Handling, cacheRejections', () => {
+    it('should not cache rejections by default', async () => {
+      const fn = createAsyncFn({
+        returnValue: { value: 'value', access: { type: 'incoming-payment' } },
+        shouldReject: true,
+        timeout: 500,
+      });
+      const dedupedFn = deduplicatorService.dedupe(fn);
+
+      // fn will reject immediately with error
+      const result1 = dedupedFn(1, { key: 'value' });
+      // wait for result1 to finish execution and reject
+      await expect(result1).rejects.toThrow('Test error');
+      // deduplicator cache is updated with the rejected result from the first call,
+      // call will not use the cache and will execute the original function fn(1, 2) again
+      const result2 = dedupedFn(1, { key: 'value' });
+      await expect(result2).rejects.toThrow('Test error');
+      expect(fn).toHaveBeenCalledTimes(2);
+    });
+
+    it('should cache and reuse pending promises when cacheRejections is false by default', async () => {
+      const fn = createAsyncFn({
+        returnValue: { value: 'value', access: { type: 'outgoing-payment' } },
+        shouldReject: true,
+        timeout: 500,
+      });
+      const dedupedFn = deduplicatorService.dedupe(fn);
+      const result1 = dedupedFn(1, 2);
+      // at this point, result1 promise is still pending,
+      // cache will return and reuse the same pending promise to result2
+      const result2 = dedupedFn(1, 2);
+      jest.runAllTimers();
+
+      await expect(result1).rejects.toThrow('Test error');
+      await expect(result2).rejects.toThrow('Test error');
+      expect(fn).toHaveBeenCalledTimes(1);
+    });
+
+    it('should cache rejections when cacheRejections is true', async () => {
+      const fn = createAsyncFn({
+        returnValue: { value: 'value', access: { type: 'incoming-payment' } },
+        shouldReject: true,
+        timeout: 500,
+      });
+      const dedupedFn = deduplicatorService.dedupe(fn, {
+        cacheRejections: true,
+      });
+
+      await expect(dedupedFn()).rejects.toThrow('Test error');
+      await expect(dedupedFn()).rejects.toThrow('Test error');
```

Comment:
Let's add an assertion that rejection of first call `toBe` equal to result of second call.